runners/spark/src/main/java/org/apache/beam/runners/spark/metrics/SparkMetricsContainer.java

```diff
@@ -0,0 +1,288 @@
+
+  private transient volatile LoadingCache<String, MetricsContainer> metricsContainers;
+
+  private final Map<MetricKey, MetricAggregator<?>> metrics = new HashMap<>();
+
+  SparkMetricsContainer() {}
+
+  public static Accumulator<SparkMetricsContainer> getAccumulator(JavaSparkContext jsc) {
+    return MetricsAccumulator.getInstance(jsc);
+  }
+
+  public MetricsContainer getContainer(String stepName) {
+    if (metricsContainers == null) {
+      synchronized (this) {
+        if (metricsContainers == null) {
+          metricsContainers = CacheBuilder.<String, SparkMetricsContainer>newBuilder()
+              .build(new MetricsContainerCacheLoader());
+        }
+      }
+    }
+    try {
+      return metricsContainers.get(stepName);
+    } catch (ExecutionException e) {
+      LOG.error("Error while creating metrics container", e);
+      return null;
+    }
+  }
+
+  Collection<CounterAggregator> getCounters() {
+    return
+        FluentIterable
+            .from(metrics.values())
+            .filter(IS_COUNTER)
+            .transform(TO_COUNTER)
+            .toList();
+  }
+
+  private static final Predicate<MetricAggregator<?>> IS_COUNTER =
+      new Predicate<MetricAggregator<?>>() {
+        @Override
+        public boolean apply(MetricAggregator<?> input) {
+          return (input instanceof CounterAggregator);
+        }
+      };
+
+  private static final Function<MetricAggregator<?>, CounterAggregator> TO_COUNTER =
+      new Function<MetricAggregator<?>,
+          CounterAggregator>() {
+        @Override
+        public CounterAggregator apply(MetricAggregator<?> metricAggregator) {
+          return (CounterAggregator) metricAggregator;
+        }
+      };
+
+  Collection<DistributionAggregator> getDistributions() {
+    return
+        FluentIterable
+            .from(metrics.values())
+            .filter(IS_DISTRIBUTION)
+            .transform(TO_DISTRIBUTION)
+            .toList();
+  }
+
+  private static final Predicate<MetricAggregator<?>> IS_DISTRIBUTION =
+      new Predicate<MetricAggregator<?>>() {
+        @Override
+        public boolean apply(MetricAggregator<?> input) {
+          return (input instanceof DistributionAggregator);
+        }
+      };
+
+  private static final Function<MetricAggregator<?>, DistributionAggregator> TO_DISTRIBUTION =
+      new Function<MetricAggregator<?>, DistributionAggregator>() {
+        @Override
+        public DistributionAggregator apply(MetricAggregator<?> metricAggregator) {
+          return (DistributionAggregator) metricAggregator;
+        }
+      };
+
+  SparkMetricsContainer merge(SparkMetricsContainer other) {
+    return
+        new SparkMetricsContainer()
+            .updated(this.getAggregators())
+            .updated(other.getAggregators());
+  }
+
+  private Collection<MetricAggregator<?>> getAggregators() {
+    return metrics.values();
+  }
+
+  private void writeObject(ObjectOutputStream out) throws IOException {
+    materialize();
```

Comment:
Can add a comment telling about how you use Java's serialization hook here to materialize the metrics before sending them back to the driver.